src/error.rs

```diff
@@ -1,26 +1,107 @@
+//! Enumerates all errors that this crate may return.
+//!
+//! [Error] is the top level error enum.
+//! Most consumers should only need to interact with this type.
+//! This is also where more generic errors such as IO errors are placed,
+//! whereas the more specific errors ([Connection] and [Protocol]) are
+//! related to logic.
+//!
+//! [Connect] describes errors specific to the connection logic, for example
+//! version mismatches or an invalid URL.
+//!
+//! [Protocol] describes lower-level errors relating to communication
+//! with the faktory server. Typically, [Protocol] errors are the result
+//! of the server sending a response this client did not expect.
+
+use thiserror::Error;
+
+/// The set of observable errors when interacting with a Faktory server.
+#[derive(Debug, Error)]
+#[non_exhaustive]
+pub enum Error {
+    /// The connection to the server, or one of its prerequisites, failed.
+    #[error("connection")]
+    Connect(#[from] Connect),
+
+    /// Underlying IO layer errors.
+    ///
+    /// These are overwhelmingly network communication errors on the socket connection to the server.
+    #[error("underlying IO")]
+    IO(#[from] std::io::Error),
+
+    /// Application-level errors.
+    ///
+    /// These generally indicate a mismatch between what the client expects and what the server expects.
+    #[error("protocol")]
+    Protocol(#[from] Protocol),
+
+    /// Faktory payloads are JSON encoded.
+    ///
+    /// This error is one that was encountered when attempting to deserialize a response from the server.
+    /// These generally indicate a mismatch between what the client expects and what the server provided.
+    #[error("deserialize payload")]
+    DeserializePayload(#[from] serde_json::Error),
```

Comment:
I think we should make this `#[source]` — it's not clear to me that we want to commit to always having a `From` impl for `serde_json::Error`. It will mean that we have to `.map_err(Error::DeserializePayload)` in one place or another, but I think that's okay.

Reply:
This required a fair amount of mapping (leading to the creation of `error::wrap_serde_io`), but I think was definitely a good change if for no other reason than that it highlighted that this error shouldn't imply it's only _de_serialization!